File: py/math/14.py
## 有了广度优先搜索，可以知道某个用户的一度、二度、三度等好友。
### ？给定两个用户，如何确定他们之间是几度好友？
### 双向广度优先搜索
##> 1. 从a出发，进行广度优先搜索，记录a的所有一度好友a1，看b是否出现在集合a1中。
##> 2. 如果没有，就搜索记录b的所有一度好友，看a是否出现在b1中。依次 a2（看b、b1是否有交集）
##> 3. 假设c在这个交集中，则a到c加b到c就是最短通路长。

import queue
## .java里的两个预留方法
  ###
  #user_nodes: 用户节点网络
  #que：某一层用户节点 即第几度好友
  #visited: 已访问的所有用户节点
  ### 如果把 user_id_a 看作圆心，它的一度好友看作第一层节点，二度好友看作第二层节点 .... ，que 队列只保留某一层的节点即可，visited 仍保存所有访问过的节点
def get_next_degree_friend(user_nodes, que, visited):
  que_return = queue.Queue() # 只保存某个用户的第几度好友
  visited_return = set() # 保存从某个用户开始到第几度好友
  while not que.empty():
    current_user_id = que.get()
    if user_nodes[current_user_id] is None:
      continue
    for friend_id in user_nodes[current_user_id].friends:
      if user_nodes[friend_id] is None:
        continue
      if friend_id in visited:
        continue
      que_return.put(friend_id)
      visited_return.add(friend_id)
  return que_return, visited_return

def has_overlap(visited_a, visited_b):
  # 两个 set的交集
  return len(visited_a & visited_b) > 0

if __name__ == "__main__":
  user_nodes_list = set_user_relation(10, 20)
  for i in range(user_nodes_list):
    print('用户 %s 的好友：%s' % (user_nodes_list[i].user_id, user_nodes_list[i].friends))
  print('----双向广度优先搜索----')
  print('两个用户节点1 和 2 之间最短路径长度：', bi_bfs(user_nodes_list, 1, 2))
  

File: py/math/14.java
/**
 * 通过双向广度优先搜索，查找两人之间最短通路的长度
 * user_nodes用户结点；user_id_a用户a的ID；user_id_b用户b的ID
 */
public static int bi_bfs(Node[] user_nodes, int user_id_a, int user_id_b) {
  if (user_id_a > user_nodes.length || user_id_b > user_nodes.length) return -1;
  if (user_id_a == user_id_b) return 0;
  Queue<Integer> queue_a = new LinkedList<Integer>(); // 队列a，用于从用户a出发的广度优先搜索
  Queue<Integer> queue_b = new LinkedList<Integer>(); // 队列b，用于从用户b出发的广度优先搜索

  queue_a.offer(user_id_a); // 放入初始结点
  HashSet<Integer> visited_a = new HashSet<>(); // 存放已经被访问过的结点，防止回路
  visited_a.add(user_id_a);

  queue_b.offer(user_id_b);
  HashSet<Integer> visited_b = new HashSet<>();
  visited_b.add(user_id_b);

  // 各自从两个方向出发，每次广度优先搜索一度，并查找是不是存在重叠好友
  int degree_a = 0, degree_b = 0, max_degree = 20;  // max_degree 防止两者之前不存在通路的情况
  while ((degree_a + degree_b) < max_degree) {
    degree_a ++;
    // 沿着a出发的方向，继续广度优先搜索degree + 1的好友
    getNextDegreeFriend(user_id_a, user_nodes, queue_a, visited_a, degree_a);
    // 判断到目前为止，被发现的a的好友，和被发现的b的好友，两个集合是否存在交集
    if (hasOverlap(visited_a, visited_b)) return (degree_a + degree_b);

    degree_b ++;
    getNextDegreeFriend(user_id_b, user_nodes, queue_b, visited_b, degree_b);
    if (hasOverlap(visited_b, visited_b)) return (degree_a + degree_b);
  }

  return -1;
}